File: programmers/스킬트리.cpp
#include <string>
#include <vector>

using namespace std;

int solution(string skill, vector<string> skill_trees) {
    int answer = 0;
    
    for(int i=0; i<skill_trees.size();i++){
        bool skillCheck=true;
        int flag=0;
        for(int j=0; j<skill_trees[i].size();j++){
            bool check=false;
            if(skill[flag]==skill_trees[i][j]){
                flag++;
                continue;
            }
            for(int k=flag; k<skill.size(); k++){
                if(skill[k]==skill_trees[i][j]){
                    check=true;
                    break;
                }
            }
            if(check){
                skillCheck=false;
                break;
            }
        }
        if(skillCheck)
            answer++;
    }
    return answer;
}

File: programmers/스킬트리.java
import java.util.*;
class Solution {

    public int solution(String skill, String[] skill_trees) {
        int answer = 0;
        Boolean[] alphabet= new Boolean[26];
        for(int i=0; i<skill_trees.length; i++){
             Arrays.fill(alphabet,false);
            for(int j=0; j<skill.length(); j++){
               alphabet[skill.charAt(j)-'A']=true;
            }
            Boolean check=true;
            int flag=0;
            for(int j=0; j<skill_trees[i].length(); j++){
                if(alphabet[skill_trees[i].charAt(j)-'A']){
                    if(skill_trees[i].charAt(j)==skill.charAt(flag)){
                        flag++;
                        alphabet[skill_trees[i].charAt(j)-'A']=false;
                    }
                    else{
                        check=false;
                        break;
                    }
                }
            }
            if(check){
                 answer++;
            }
               
        }
        return answer;
    }
}